cpu/atmega_common/periph/i2c.c

```diff
@@ -0,0 +1,347 @@
+/*
+ * Copyright (C) 2017 Hamburg University of Applied Sciences, Dimitri Nahm
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @addtogroup  driver_periph
+ * @{
+ *
+ * @file
+ * @brief       Low-level I2C driver implementation fot atmega common
+ *
+ * @note This implementation only implements the 7-bit addressing mode.
+ *
+ * @author      Dimitri Nahm <dimitri.nahm@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+
+#include "cpu.h"
+#include "mutex.h"
+#include "periph/i2c.h"
+#include "periph_conf.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define MT_START 0x08
+#define MT_ADDRESS_ACK 0x18
+#define MT_DATA_ACK 0x28
+#define MR_ADDRESS_ACK 0x40
+
+/* guard file in case no I2C device is defined */
+#if I2C_NUMOF
+
+/* static function definitions */
+static int _start(uint8_t address, uint8_t rw_flag);
+static int _write(const uint8_t *data, int length);
+static void _stop(void);
+
+/**
+ * @brief Array holding one pre-initialized mutex for each I2C device
+ */
+static mutex_t locks[] =  {
+#if I2C_0_EN
+    [I2C_0] = MUTEX_INIT,
+#endif
+};
+
+int i2c_init_master(i2c_t dev, i2c_speed_t speed)
+{
+    /* TWI Bit Rate Register - division factor for the bit rate generator*/
+    int twibrr;
+
+    /* check if the line is valid */
+    if (dev >= I2C_NUMOF) {
+        return -1;
+    }
+
+    /* calculate speed configuration */
+    switch (speed) {
+        case I2C_SPEED_NORMAL:
+            twibrr = ((CLOCK_CORECLOCK/100000)-16)/2;
+            break;
+
+        case I2C_SPEED_FAST:
+            twibrr = ((CLOCK_CORECLOCK/400000)-16)/2;
+            break;
+
+        default:
+            return -2;
+    }
+
+    /* set pull-up on SCL and SDA */
+    #if defined (CPU_ATMEGA2560) || defined (CPU_ATMEGA1281)
+        PORTD |= (1 << PORTD0) | (1 << PORTD1);
+    #endif
+    #ifdef CPU_ATMEGA328P
+        PORTC |= (1 << PORTC4) | (1 << PORTC5);
+    #endif
```

Comment:
also: this should be configurable in the boards config (periph_conf.h). So I think the code here should look something like:
```c
#if I2C_PULL_ENABLE    // -> this define is [optinally] defined in a board's periph_conf.h
    I2C_PORT_REG |= (I2C_PIN_MASK);    // -> these are defined in the CPU's periph_cpu.h
#endif
```